pkg/envoy/lds/ingress.go

```diff
@@ -57,27 +65,43 @@ func newIngressFilterChain(cfg configurator.Configurator, svc service.MeshServic
 	}
 }
 
-func getIngressFilterChains(svc service.MeshService, cfg configurator.Configurator) []*xds_listener.FilterChain {
+func (lb *listenerBuilder) getIngressFilterChains(svc service.MeshService) []*xds_listener.FilterChain {
 	var ingressFilterChains []*xds_listener.FilterChain
 
-	if cfg.UseHTTPSIngress() {
-		// Filter chain with SNI matching enabled for HTTPS clients that set the SNI
-		ingressFilterChainWithSNI := newIngressFilterChain(cfg, svc)
-		ingressFilterChainWithSNI.Name = inboundIngressHTTPSFilterChain
-		ingressFilterChainWithSNI.FilterChainMatch.ServerNames = []string{svc.ServerName()}
-		ingressFilterChains = append(ingressFilterChains, ingressFilterChainWithSNI)
+	protocolToPortMap, err := lb.meshCatalog.GetPortToProtocolMappingForService(svc)
+	if err != nil {
+		log.Error().Err(err).Msgf("Error retrieving port to protocol mapping for service %s", svc)
+		return ingressFilterChains
 	}
 
-	// Filter chain without SNI matching enabled for HTTP clients and HTTPS clients that don't set the SNI
-	ingressFilterChainWithoutSNI := newIngressFilterChain(cfg, svc)
-	ingressFilterChainWithoutSNI.Name = inboundIngressHTTPFilterChain
-	ingressFilterChains = append(ingressFilterChains, ingressFilterChainWithoutSNI)
+	// Create protocol specific inbound filter chains per port to handle different ports serving different protocols
+	for port, appProtocol := range protocolToPortMap {
+		switch appProtocol {
+		case httpAppProtocol:
+			// Ingress filter chain for HTTP port
+			if lb.cfg.UseHTTPSIngress() {
+				// Filter chain with SNI matching enabled for HTTPS clients that set the SNI
+				ingressFilterChainWithSNI := newIngressHTTPFilterChain(lb.cfg, svc, port)
+				ingressFilterChainWithSNI.Name = fmt.Sprintf("%s:%d", inboundIngressHTTPSFilterChain, port)
+				ingressFilterChainWithSNI.FilterChainMatch.ServerNames = []string{svc.ServerName()}
+				ingressFilterChains = append(ingressFilterChains, ingressFilterChainWithSNI)
+			}
+
+			// Filter chain without SNI matching enabled for HTTP clients and HTTPS clients that don't set the SNI
+			ingressFilterChainWithoutSNI := newIngressHTTPFilterChain(lb.cfg, svc, port)
+			ingressFilterChainWithoutSNI.Name = fmt.Sprintf("%s:%d", inboundIngressNonSNIFilterChain, port)
+			ingressFilterChains = append(ingressFilterChains, ingressFilterChainWithoutSNI)
+
+		default:
+			log.Error().Msgf("Cannot build inbound filter chain, unsupported protocol %s for proxy:port %s:%d", appProtocol, svc, port)
+		}
+	}
 
 	return ingressFilterChains
 }
 
-func getIngressTransportSocket(cfg configurator.Configurator, marshalledDownstreamTLSContext *any.Any) *xds_core.TransportSocket {
-	if cfg.UseHTTPSIngress() {
+func getIngressTransportSocket(forHTTPS bool, marshalledDownstreamTLSContext *any.Any) *xds_core.TransportSocket {
+	if forHTTPS {
```

Comment:
IMO, I'd put the condition outside. having a parametre to call a function and have it do nothing, better not call it from the start.

it throws dirt on the creation of the struct though, up to you.

Reply:
I'll leave this as is for now instead of refactoring code outside the scope of this PR. The existing code has generic functions that branch out based on `cfg.UseHTTPSIngress()`, so to do this differently the existing code (not changed in this PR) needs refactoring.